src/systemd-sonic-generator/systemd-sonic-generator.c

```diff
@@ -0,0 +1,264 @@
+#include <assert.h>
+#include <stdio.h>
+#include <unistd.h>
+#include <errno.h>
+#include <string.h>
+#include <stdlib.h>
+#include <stdbool.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <linux/limits.h>
+
+#define MAX_NUM_TARGETS 5
+#define MAX_NUM_UNITS 128
+
+static const char* UNIT_FILE_PREFIX = "/etc/systemd/system/";
+static const char* CONFIG_FILE = "/etc/sonic/generated_services.conf";
+
+
+void strip_trailing_newline(char* str) {
+    /***
+    Strips trailing newline from a string if it exists
+    ***/
+
+    size_t l = strlen(str);
+    if (l > 0 && str[l-1] == '\n')
+        str[l-1] = '\0';
+}
+
+
+static int get_target_lines(char* unit_file, char* target_lines[]) {
+    /***
+    Gets installation information for a given unit file
+
+    Returns lines in the [Install] section of a unit file
+    ***/
+    FILE *fp;
+    char *line = NULL;
+    size_t len = 0;
+    ssize_t read;
+    bool found_install;
+    int num_target_lines;
+
+
+    fp = fopen(unit_file, "r");
+
+    if (fp == NULL) {
+        exit(EXIT_FAILURE);
+    }
+
+    found_install = false;
+    num_target_lines = 0;
+
+    while ((read = getline(&line, &len, fp)) != -1 ) {
+        // Assumes that [Install] is the last section of the unit file
+        if (strstr(line, "[Install]") != NULL) {
+             found_install = true;
+        }
+        else if (found_install) {
+            target_lines[num_target_lines] = strdup(line);
+            num_target_lines++;
```

Comment:
please check that num_target_lines not more than you have reserved for target_lines